### PR TITLE
Rework GW and BSE method

### PR DESCRIPTION
**Pull request by JosePizarro3**
@ladinesa I worked a bit more on the GW and BSE parsing for all the electronic codes (OCEAN, exciting, FHIaims, VASP, ABINIT).

As you can maybe see in the corresponding nomad MR, now there is a base class called `ExcitedStateMethodology`. This contains the shared method parameters amongst `Screening`, `GW` and `BSE`. Each of these methods can contain extra quantities or sub-sections (e.g., `dielectric_infinity` quantity for `Screening`, or `core` subsection for `BSE`).

Please, also check the new syntax, as I think now it is a bit more clean. This also goes with the use of `m_add_sub_section`.

I will also:
- [x] Adapt and improve the testing to the new metainfo.
- [x] Adapt `utils.py` to use `m_` operations.

### CODE REVIEW

File: electronicparsers/abinit/parser.py [hunk 1]
```diff
@@ -1084,32 +1086,39 @@ def parse_gw(self):
...
-        # Frequency grid
+        # FreqMesh
         if self.out_parser.get('screening_dataset'):
             if self.out_parser.get('screening_dataset').get('frequencies'):
                 freqs = self.out_parser.get('screening_dataset').get('frequencies').get('values') * ureg.eV
-                sec_freq_grid = sec_gw.m_create(FreqMesh)
-                sec_freq_grid.n_points = len(freqs)
-                sec_freq_grid.values = freqs[:, 0] + freqs[:, 1] * 1j
+                values = freqs[:, 0] + freqs[:, 1] * 1j
+                sec_freq_mesh = FreqMesh(
```
Comment (ladinesa) [inline electronicparsers/abinit/parser.py]: no line break

File: electronicparsers/abinit/parser.py [hunk 2]
```diff
@@ -1084,32 +1086,39 @@ def parse_gw(self):
...
+                    n_points=len(freqs),
+                    values=values)
         else:
             freq_plasma = self.out_parser.get('gw_dataset').get('omega_plasma', 0.0)
```
Comment (ladinesa) [inline electronicparsers/abinit/parser.py]: error handling

File: electronicparsers/abinit/parser.py [hunk 3]
```diff
@@ -1084,32 +1086,39 @@ def parse_gw(self):
...
-            sec_freq_grid = sec_gw.m_create(FreqMesh)
-            sec_freq_grid.n_points = 2
-            sec_freq_grid.values = [0.0, freq_plasma * 1j]
+            values = [0.0, freq_plasma * 1j]
+            sec_freq_mesh = FreqMesh(
```
Comment (ladinesa) [inline electronicparsers/abinit/parser.py]: no line break

File: electronicparsers/abinit/parser.py [hunk 4]
```diff
@@ -1084,32 +1086,39 @@ def parse_gw(self):
...
+                n_points=2,
+                values=values)
+        sec_gw.m_add_sub_section(GW.frequency_mesh, sec_freq_mesh)
+        # Screening
+        occ_scr = self.out_parser.get_input_var('occ', 3, [])
+        n_scr = len(occ_scr)
+        n_occ_scr = len([occ_scr[i] for i in range(n_scr) if occ_scr[i] > 0.0])
+        n_empty_scr = n_scr - n_occ_scr
+        sec_screening = Screening(
```
Comment (ladinesa) [inline electronicparsers/abinit/parser.py]: no line break

File: electronicparsers/abinit/parser.py [hunk 5]
```diff
@@ -1084,32 +1086,39 @@ def parse_gw(self):
...
-            sec_freq_grid.values = [0.0, freq_plasma * 1j]
+            values = [0.0, freq_plasma * 1j]
+            sec_freq_mesh = FreqMesh(
+                n_points=2,
+                values=values)
+        sec_gw.m_add_sub_section(GW.frequency_mesh, sec_freq_mesh)
+        # Screening
+        occ_scr = self.out_parser.get_input_var('occ', 3, [])
+        n_scr = len(occ_scr)
+        n_occ_scr = len([occ_scr[i] for i in range(n_scr) if occ_scr[i] > 0.0])
```
Comment (ladinesa) [inline electronicparsers/abinit/parser.py]: len([occ for occ in occ_scr if occ > 0])

File: electronicparsers/exciting/parser.py [hunk 6]
```diff
@@ -1780,25 +1785,40 @@ def parse_xs(self):
...
-        sec_k_mesh.grid = sec_run.method[0].get('x_exciting_xs_ngridk')
+        sec_k_mesh.grid = sec_run.method[-1].get('x_exciting_xs_ngridk')
+        # QMesh
+        sec_q_mesh = KMesh(grid=sec_run.method[-1].get('x_exciting_xs_ngridq'))
+        sec_bse.m_add_sub_section(BSE.q_mesh, sec_q_mesh)
+        # FreqMesh
+        n_freqs = sec_run.method[-1].get('x_exciting_xs_energywindow_points')
+        freqs = sec_run.method[-1].get('x_exciting_xs_energywindow_values')
+        values = [freqs[0] + i * (freqs[-1] - freqs[0]) / n_freqs for i in range(n_freqs)]
+        sec_freq_mesh = FreqMesh(
```
Comment (ladinesa) [inline electronicparsers/exciting/parser.py]: no line break

File: electronicparsers/ocean/parser.py [hunk 7]
```diff
@@ -117,59 +117,76 @@ def parse_method(self, archive):
...
-            sec_haydock = sec_bse_ocean.m_create(x_ocean_core_haydock_parameters)
-            sec_haydock.x_ocean_converge_spacing = self.data['bse']['core']['haydock']['converge'].get('spacing')
-            sec_haydock.x_ocean_converge_thresh = self.data['bse']['core']['haydock']['converge'].get('thresh')
-            sec_haydock.x_ocean_niter = self.data['bse']['core']['haydock'].get('niter')
-        elif sec_core.solver == 'gmres':
-            sec_gmres = sec_bse_ocean.m_create(x_ocean_core_gmres_parameters)
-            gmres_keys = ['echamp', 'elist', 'erange', 'estyle', 'ffff', 'gprc', 'nloop']
-            for key in gmres_keys:
-                setattr(sec_gmres, f'x_ocean_{key}', self.data['bse']['core']['gmres'].get(key))
+        try:
```
Comment (ladinesa) [inline electronicparsers/ocean/parser.py]: this is way too much of a try block. can you check exactly where a possible error can occur

File: electronicparsers/vasp/parser.py [hunk 8]
```diff
@@ -1361,24 +1361,17 @@ def parse_gw(self):
...
-        # Frequency grid
-        n_omega = response_functions.get('NOMEGA', 100)
-        sec_freq_grid = sec_gw.m_create(FreqMesh)
-        sec_freq_grid.n_points = n_omega
-
-        # Other parameters
-        n_bands = self.parser.get_incar().get('NBANDS')
-        sec_gw.n_states_self_energy = n_bands
+        # FreqMesh
+        sec_freq_mesh = FreqMesh(
```
Comment (ladinesa) [inline electronicparsers/vasp/parser.py]: no line break

File: electronicparsers/abinit/parser.py (hunk 1 above)
Comment (JosePizarro3) [reply electronicparsers/abinit/parser.py]: done

File: electronicparsers/abinit/parser.py (hunk 2 above)
Comment (JosePizarro3) [reply electronicparsers/abinit/parser.py]: done

File: electronicparsers/abinit/parser.py (hunk 3 above)
Comment (JosePizarro3) [reply electronicparsers/abinit/parser.py]: done

File: electronicparsers/abinit/parser.py (hunk 4 above)
Comment (JosePizarro3) [reply electronicparsers/abinit/parser.py]: done

File: electronicparsers/abinit/parser.py (hunk 5 above)
Comment (JosePizarro3) [reply electronicparsers/abinit/parser.py]: done

File: electronicparsers/exciting/parser.py (hunk 6 above)
Comment (JosePizarro3) [reply electronicparsers/exciting/parser.py]: done

File: electronicparsers/ocean/parser.py (hunk 7 above)
Comment (JosePizarro3) [reply electronicparsers/ocean/parser.py]: you are right, I simply removed this try/except, as the error handling there seems to be good

File: electronicparsers/vasp/parser.py (hunk 8 above)
Comment (JosePizarro3) [reply electronicparsers/vasp/parser.py]: done

**Comment by JosePizarro3** [comment]
Ok, I implemented your changes, as well as working on testing and utils.py.

Can you please check `utils.py`? I also tested it but just to make sure.

File: electronicparsers/utils/utils.py [hunk 9]
```diff
@@ -120,46 +120,58 @@ def parse_gw_workflow(self, gw_archive, gw_workflow_archive):
...
         workflow.results.band_structure_dft = bs_dft
         workflow.results.band_structure_gw = bs_gw
-        # inputs and outputs
+
+        # Inputs and Outputs
         input_structure = extract_section(self.archive, 'run/system')
         dft_calculation = extract_section(self.archive, 'run/calculation')
         gw_calculation = extract_section(gw_archive, 'run/calculation')
-        workflow.inputs.append(Link(name='Workflow parameters', section=workflow.method))
+        # workflow.m_add_sub_section(
```
Comment (ladinesa) [inline electronicparsers/utils/utils.py]: remove comment

Comment (JosePizarro3) [reply electronicparsers/utils/utils.py]: forgot it, already done :smile:

File: electronicparsers/utils/utils.py [hunk 10]
```diff
@@ -168,41 +180,50 @@ def parse_photon_workflow(self):
...
+        input_method = extract_section(self.archive, 'run/method')
+        workflow.m_add_sub_section(
+            PhotonPolarization.inputs, Link(name='Input BSE methodology', section=input_method))
+
+        # Outputs
         spectra = []
-        outputs = []
         for path in self._child_archives.keys():
             archive = self._child_archives.get(path)
             index = list(self._child_archives.keys()).index(path)
```
Comment (ladinesa) [inline electronicparsers/utils/utils.py]: define index list out of loop

File: electronicparsers/utils/utils.py [hunk 11]
```diff
@@ -226,40 +247,53 @@ def extract_polarization_outputs():
...
-                    archive.workflow2.inputs.append(Link(name=output_dft_name, section=dft_calculation))
-                    archive.workflow2.tasks[0].inputs.append(Link(name=output_dft_name, section=dft_calculation))
+        polarization_calculations = extract_polarization_outputs()
+        workflow.m_add_sub_section(
+            ParticleHoleExcitations.inputs,
+            Link(name='Particle-Hole excitations workflow parameters', section=workflow.method))
+        if input_structure:
+            workflow.m_add_sub_section(
+                ParticleHoleExcitations.inputs, Link(name='Input structure', section=input_structure))
+        for polarizations in polarization_calculations:
```
Comment (ladinesa) [inline electronicparsers/utils/utils.py]: enumerate

File: electronicparsers/utils/utils.py [hunk 12]
```diff
@@ -226,40 +247,53 @@ def extract_polarization_outputs():
...
             if dft_calculation:
-                self.archive.workflow2.outputs.append(Link(name=output_dft_name, section=dft_calculation))
-            workflow.tasks.append(task)
-        # PhotonPolarization task
+                task.outputs = [Link(name='Output DFT calculation', section=dft_calculation)]
+            workflow.m_add_sub_section(ParticleHoleExcitations.tasks, task)
+
+        # Spectra task
         for xs_archive in xs_archives:
             if xs_archive.workflow2:
```
Comment (ladinesa) [inline electronicparsers/utils/utils.py]: im not a big fan of indents so do if not xs_archive.workflow2: continue

**Comment by ladinesa** [comment]
BeyondDFT is getting bigger, do you think we need to split it up then just do multiple inheritances if we plan to add more later?

File: electronicparsers/utils/utils.py (hunk 12 above)
Comment (JosePizarro3) [reply electronicparsers/utils/utils.py]: done

File: electronicparsers/utils/utils.py (hunk 11 above)
Comment (JosePizarro3) [reply electronicparsers/utils/utils.py]: done

File: electronicparsers/utils/utils.py (hunk 10 above)
Comment (JosePizarro3) [reply electronicparsers/utils/utils.py]: done with enumerate

File: electronicparsers/utils/utils.py (hunk 11 above)
Comment (JosePizarro3) [reply electronicparsers/utils/utils.py]: here and below again for `# Spectra task`

**Comment by JosePizarro3** [comment]
> BeyondDFT is getting bigger, do you think we need to split it up then just do multiple inheritances if we plan to add more later?

Yeah, I think it makes sense, but I am not sure how to handle this. Maybe we can meet at some point next or in 2 weeks to tackle it? Whenever you want :slightly_smiling_face:

File: electronicparsers/exciting/parser.py [hunk 13]
```diff
@@ -1780,25 +1785,38 @@ def parse_xs(self):
         # Code-specific
         self.parse_file('input.xml', sec_method, self._xs_info_file)
 
+        # BSE
+        sec_bse = sec_method.m_create(BSE)
+        sec_bse.type = sec_run.method[-1].get('x_exciting_xs_bse_type')
+        sec_bse.n_empty_states = sec_run.method[-1].get('x_exciting_xs_number_of_empty_states')
+        sec_bse.broadening = sec_run.method[-1].get('x_exciting_xs_broadening') * ureg.hartree
```
Comment (ladinesa) [inline electronicparsers/exciting/parser.py]: error handling

File: electronicparsers/ocean/parser.py [hunk 14]
```diff
@@ -117,24 +117,44 @@ def parse_method(self, archive):
         if sec_run.m_xpath('method[0].photon'):
             sec_method.starting_method_ref = sec_run.method[0]
 
-        # KMesh section
+        # BSE
+        sec_bse = sec_method.m_create(BSE)
+        bse_data = self.data.get('bse', {})
+        sec_bse.type = self._type_bse_map[bse_data.get('val', {}).get('solver')]
+        sec_bse.n_states = bse_data.get('nbands', 1)
+        sec_bse.broadening = bse_data.get('val', {}).get('broaden') * ureg.eV
```
Comment (ladinesa) [inline electronicparsers/ocean/parser.py]: error handling

File: electronicparsers/ocean/parser.py [hunk 15]
```diff
@@ -117,24 +117,44 @@ def parse_method(self, archive):
...
-        sec_bse.dielectric_infinity = self.data['structure'].get('epsilon')
-        sec_bse.n_empty_states_screening = self.data['screen'].get('nbands')
-        sec_bse.k_mesh_screening = KMesh(grid=self.data['screen'].get('kmesh'))
+        # Core-Hole (either K=1s or L23=2p depenging on the first edge found)
+        bse_core_data = bse_data.get('core')
+        if bse_core_data:
+            sec_core_hole = CoreHole(
+                mode=self.mode_bse[bse_core_data.get('strength')],
+                solver=self._type_bse_map[bse_core_data.get('solver')],
+                broadening=bse_core_data.get('broaden') * ureg.eV)
```
Comment (ladinesa) [inline electronicparsers/ocean/parser.py]: error handling

**Comment by ladinesa** [comment]
we should start putting variable types.we start with utils.py but we can do it in a separate issue.

File: electronicparsers/abinit/parser.py [hunk 16]
```diff
@@ -1084,32 +1086,34 @@ def parse_gw(self):
                 sec_gw.analytical_continuation = 'ppm_FaridEngel'
             else:
                 sec_gw.analytical_continuation = 'contour_deformation'
-        # Frequency grid
+        # FrequencyMesh
         if self.out_parser.get('screening_dataset'):
             if self.out_parser.get('screening_dataset').get('frequencies'):
                 freqs = self.out_parser.get('screening_dataset').get('frequencies').get('values') * ureg.eV
```
Comment (ladinesa) [inline electronicparsers/abinit/parser.py]: error handling. The multiple if here does not suit me can we do it in one line? 
```
value = self.out_parser.get('screening_dataset', {}).get('frequency', {}).get('values')
if value is not None:
    ....
```

File: electronicparsers/exciting/parser.py (hunk 13 above)
Comment (JosePizarro3) [reply electronicparsers/exciting/parser.py]: my bad. This quantities are always set by the input, so I change the `get('...')` to a simple call to the quantity inside the section

File: electronicparsers/ocean/parser.py (hunk 14 above)
Comment (JosePizarro3) [reply electronicparsers/ocean/parser.py]: done, set to 0.1 by default

File: electronicparsers/ocean/parser.py (hunk 15 above)
Comment (JosePizarro3) [reply electronicparsers/ocean/parser.py]: done, set to 0.1 by default

**Comment by JosePizarro3** [comment]
> we should start putting variable types.we start with utils.py but we can do it in a separate issue.

how do you mean?

**Comment by ladinesa** [comment]
> > we should start putting variable types.we start with utils.py but we can do it in a separate issue.
> 
> how do you mean?

like
```
def parse_gw_workflow(self, gw_archive: EntryArchive, gw_workflow_archive: EntryArchive)
```

File: electronicparsers/abinit/parser.py [hunk 17]
```diff
@@ -1084,32 +1086,34 @@ def parse_gw(self):
...
-            sec_freq_grid.values = [0.0, freq_plasma * 1j]
+            freq_plasma = self.out_parser.get('gw_dataset', {}).get('omega_plasma', 0.0)
+            values = [0.0, freq_plasma * 1j]
+            sec_freq_mesh = FrequencyMesh(n_points=2, values=values)
+        sec_gw.m_add_sub_section(GW.frequency_mesh, sec_freq_mesh)
+        # Screening
+        occ_scr = self.out_parser.get_input_var('occ', 3, [])
+        n_scr = len(occ_scr)
+        n_occ_scr = len([occ for occ in occ_scr if occ > 0.0])
+        n_empty_scr = n_scr - n_occ_scr
```
Comment (ladinesa) [inline electronicparsers/abinit/parser.py:1104]: necessary to define n_empty_scr?

File: electronicparsers/exciting/parser.py [hunk 18]
```diff
@@ -1780,25 +1785,38 @@ def parse_xs(self):
         # Code-specific
         self.parse_file('input.xml', sec_method, self._xs_info_file)
 
+        # BSE
+        sec_bse = sec_method.m_create(BSE)
+        sec_bse.type = sec_run.method[-1].x_exciting_xs_bse_type
+        sec_bse.n_empty_states = sec_run.method[-1].x_exciting_xs_number_of_empty_states
+        sec_bse.broadening = sec_run.method[-1].x_exciting_xs_broadening * ureg.hartree
```
Comment (ladinesa) [inline electronicparsers/exciting/parser.py:1792]: is this always assigned? if not error handling

File: electronicparsers/exciting/parser.py [hunk 19]
```diff
@@ -1889,21 +1909,32 @@ def parse_gw(self):
...
+            type=sec_gw.x_exciting_freqgrid.x_exciting_fgrid,
+            n_points=n_freqs,
+            values=values,
+            smearing=smearing)
+        sec_gw.m_add_sub_section(GW.frequency_mesh, sec_freq_mesh)
+        # Screening
+        sec_screening = Screening(
+            type=sec_gw.x_exciting_scrcoul.x_exciting_scrtype,
+            n_empty_states=sec_gw.x_exciting_nempty,
+            k_mesh=sec_k_mesh.m_copy(),
```
Comment (ladinesa) [inline electronicparsers/exciting/parser.py]: Is this a reference in the schema? it appears not to me if we copy. If not reference, is this necessary?

File: electronicparsers/ocean/parser.py [hunk 20]
```diff
@@ -143,33 +163,27 @@ def parse_method(self, archive):
         screen_dicts = [
             'core_offset', 'final', 'grid']
         for key in screen_keys:
-            setattr(sec_bse_screen, f'x_ocean_{key}', self.data['screen'].get(key))
+            setattr(sec_bse_screen, f'x_ocean_{key}', screen_data.get(key))
         for keys in screen_dicts:
```
Comment (ladinesa) [inline electronicparsers/ocean/parser.py]: Very minor but, key or subkey is more appropriate?

File: electronicparsers/fhiaims/parser.py [hunk 21]
```diff
@@ -857,24 +857,29 @@ def parse_gw(self):
...
-        sec_freq_grid = sec_gw.m_create(FreqMesh)
-        sec_freq_grid.type = self.out_parser.get('freq_grid_type')
-        sec_freq_grid.n_points = self.out_parser.get('n_freq', 100)
+        # FrequencyMesh
         frequency_data = self.out_parser.get('frequency_data', None)
-        if frequency_data is not None:
-            sec_freq_grid.values = np.array(frequency_data)[:, 1] * ureg.hartree
-        # Other parameters
-        sec_gw.n_states_self_energy = self.out_parser.get('n_states_gw')
+        if frequency_data:
```
Comment (ladinesa) [inline electronicparsers/fhiaims/parser.py]: this if will not work if array. is not None is better or len(frequency_data) > 0 if you set default to [].

File: electronicparsers/exciting/parser.py (hunk 18 above)
Comment (JosePizarro3) [reply electronicparsers/exciting/parser.py:1792]: it is (to 0.01 Ha), line 1497

File: electronicparsers/abinit/parser.py (hunk 17 above)
Comment (JosePizarro3) [reply electronicparsers/abinit/parser.py:1104]: I forgot to save `n_scr` on `Screening`. I'll leave both quantities to check later which one is more useful for the GW/BSE community when searching the Archive.

File: electronicparsers/exciting/parser.py (hunk 19 above)
Comment (JosePizarro3) [reply electronicparsers/exciting/parser.py]: deleted, you are right

File: electronicparsers/ocean/parser.py (hunk 20 above)
Comment (JosePizarro3) [reply electronicparsers/ocean/parser.py]: done, both here and down for `subkey`

File: electronicparsers/fhiaims/parser.py (hunk 21 above)
Comment (JosePizarro3) [reply electronicparsers/fhiaims/parser.py]: done, thanks!

File: electronicparsers/exciting/parser.py (hunk 18 above)
Comment (ladinesa) [reply electronicparsers/exciting/parser.py:1792]: ah i see.

**Comment by JosePizarro3** [comment]
Done with this next round of changes. Thanks a lot!

Feel free to resolve if you are satisfied, and keep up on the suggestions

**Comment by JosePizarro3** [comment]
> we should start putting variable types.we start with utils.py but we can do it in a separate issue

Ok, I did it for utils.py, and will have it in mind for the future.

File: electronicparsers/abinit/parser.py (hunk 16 above)
Comment (JosePizarro3) [reply electronicparsers/abinit/parser.py]: done

**Comment by ladinesa** [comment]
> > we should start putting variable types.we start with utils.py but we can do it in a separate issue
> 
> Ok, I did it for utils.py, and will have it in mind for the future.

thanks. please make sure to run pylint pycodestyle locally to make sure we do not miss any linting errors as a result of this.

**Pull request merged**